### PR TITLE
Proxy site

### PR DESCRIPTION
**Pull request by mgiuca**
Added proxy site to as an intermediary between requester and handler.

Handlers using the polyfill will now prompt the user to be registered
locally upon loading. Requesters using the polyfill will now prompt the
user to pick a registered handler (rather than being required to
hard-code a specific handler).

### CODE REVIEW

File: polyfill/ballista-polyfill.js [hunk 1]
```diff
@@ -101,10 +111,35 @@ if (self.WorkerNavigator !== undefined) {
...
+    iframe.setAttribute('src', url);
+    iframe.style.width = '100%';
+    iframe.style.height = '100%';
+
+    chooserDiv.appendChild(iframe);
+    document.body.appendChild(chooserDiv);
+  });
+}
+
 // Establishes a connection with the polyfill handler (by embedding a page from
```
Comment (sammc) [inline polyfill/ballista-polyfill.js]: Update this.

File: polyfill/ballista-polyfill.js [hunk 2]
```diff
@@ -114,39 +149,53 @@ function sendPortToHandler(url, port) {
...
 
-  // TODO(mgiuca): document.body.removeChild(iframe), after done using it.
+  document.body.removeChild(chooserDiv);
 }
 
-// Establish a connection with the polyfill handler (by embedding a page from
-// the handler's domain in an iframe). Returns (in a promise) a MessagePort on
-// succesful connection to the handler.
-function connectToHandler(url) {
+// Prompts the user for a handler app and establish a connection with the chosen
```
Comment (sammc) [inline polyfill/ballista-polyfill.js]: establishes

File: polyfill/ballista-polyfill.js [hunk 3]
```diff
@@ -114,39 +149,53 @@ function sendPortToHandler(url, port) {
...
   return new Promise(function(resolve, reject) {
     var channel = new MessageChannel();
     channel.port1.onmessage = e => {
+      closeChooser();
+
       if (e.data.connected) {
         channel.port1.onmessage = null;
         resolve(channel.port1);
+      } else if (e.data.connected == false) {
+        // Fail silently. (This happens if the user cancels the dialog.)
```
Comment (sammc) [inline polyfill/ballista-polyfill.js]: Don't leave the promise hanging. Cancellation needs to be reported.

File: polyfill/ballista-polyfill.js [hunk 4]
```diff
@@ -354,4 +389,53 @@ function onMessageReceived(data, port) {
...
 
+// Registers a handler with the proxy server.
+function registerHandler(handler) {
+  // This channel will be used to get back a "close" signal from the proxy.
+  var channel = new MessageChannel();
+  channel.port1.onmessage = e => {
+    if (e.data.close) {
+      closeChooser();
+    } else {
+      reject(new Error("Received unexpected response from proxy."));
```
Comment (sammc) [inline polyfill/ballista-polyfill.js]: There is no promise here to reject.

File: polyfill/ballista-polyfill.js (hunk 1 above)
Comment (mgiuca) [inline polyfill/ballista-polyfill.js]: What?

File: polyfill/ballista-polyfill.js (hunk 4 above)
Comment (mgiuca) [inline polyfill/ballista-polyfill.js]: Done.

File: polyfill/ballista-polyfill.js (hunk 2 above)
Comment (mgiuca) [inline polyfill/ballista-polyfill.js]: Done.

File: polyfill/ballista-polyfill.js (hunk 1 above)
Comment (mgiuca) [inline polyfill/ballista-polyfill.js]: Done.

File: polyfill/ballista-polyfill.js (hunk 3 above)
Comment (mgiuca) [inline polyfill/ballista-polyfill.js]: Done.

**Comment by mgiuca** [comment]
OK responded to your comments.

Umm... yeah GitHub is pretty bad for code reviews (possibly because I rebased). You should look at the delta between    e633c2c and 06ec7ae.

**Comment by mgiuca** [comment]
Continuing code review on https://codereview.chromium.org/1419423003/.

**Pull request merged**